### ISSUE TITLE
Auto-commit semantics

### ISSUE THREAD

**Issue by ms7s** [dpkp/kafka-python#722]
As per https://github.com/dpkp/kafka-python/issues/292#issuecomment-169883834, messages are "consumed" as soon as they are delivered to the user from the client. Have you considered treating the messages as "consumed" only after being fully processed (i.e. when the client asks for the next message/batch?). 

So when you have a consumer that crashes due to something unexpected in a message, you fix the consumer and restart it, but the message that caused the crash is not reprocessed sometimes. Of course you can adjust the consumer offset manually and then restart the consumer, but I think it makes sense to also commit only after the message was fully processed. People may expect it works like this, as it _looks_ like it works this way when trying the default example (since it is very unlikely the program fetches the message, autocommit-timer fires, and then the program crashes), but it may be demonstrated by writing the example this way:

``` python
# To consume latest messages and auto-commit offsets
consumer = KafkaConsumer('my-topic',
                         group_id='my-group',
                         bootstrap_servers=['localhost:9092'])

try:
    for message in consumer:
        # message value and key are raw bytes -- decode if necessary!
        # e.g., for unicode: `message.value.decode('utf-8')`
        print ("\n\n%s:%d:%d: key=%s value=%s\n\n" % (message.topic, message.partition,
                                                      message.offset, message.key,
                                                      message.value))
        if message.value == b'!':
            raise ValueError('Oh snap!')
finally:
    consumer.commit()
```

(i.e. in my tests using `kafka-python==1.2.1` and broker 0.9, without the finally, the message `!` is usually reprocessed, but with the finally block, it never is)

Since Kafka documentation at http://kafka.apache.org/documentation.html#semantics says you get at-least-once message delivery semantics by default (which is technically true, as it does not speak about the message processing), I expected it would reprocess the message.

This behaviour could be implemented (if desired, since it may add a difference w.r.t the java consumer) by setting the offset to be committed at the beginning of `__next__()` instead of at the end (i.e. when the next message is requested, but before the network request, if any). In any case, I think this should be better documented since the old consumer had `task_done` and the changelog only says

> task_done [handled internally by auto-commit; or commit offsets manually]

Do you think it is worth implementing such a change? Is there any documentation regarding best practices for writing consumers and how to handle consumer restarts? Also, what happens in case of consumer rebalance?

**Comment by dpkp** [comment]
Your snippet above is manually committing offsets after failures (finally runs in both success and failure exits). Why? If you let autocommit operate normally, I believe you will find that the 'Oh snap!' message is reprocessed as you expect.

**Comment by ms7s** [comment]
Usually, yes. But the auto-commit might commit it anyway, since it is triggered by a timer.

Consider this sequence of events:

```
my code: for message in consumer:
timer:   auto-commit `message.offset + 1`
my code: fail
```

In this case, the message would not be reprocessed.

**Comment by dpkp** [comment]
There is no background processing timer. That sequence is not possible...?

**Comment by ms7s** [comment]
Ah, sorry, now I see. I misunderstood the code. There is a timer that spawns background thread in `kafka.consumer.base.Consumer`, but this is used only in the old consumer classes, not the new one. The new one has a queue of tasks to process that is executed on `KafkaClient.poll()`.

Thanks for helping me understand how the client works.

Just a question so that I know for certain, as it is not documented (not that I want to do it now) - is it safe to call `KafkaConsumer.position()` and `KafkaConsumer.topics()` from within the `for message in consumer` loop? Both these methods might call `KafkaClient.poll()` internally, couldn't they trigger the `AutoCommitTask`?

**Comment by dpkp** [comment]
Great point about `position()` or `topics()` -- any internal call to poll() could potentially interfere with autocommit.

**Comment by dpkp** [comment]
I sent a note to the kafka dev list because I believe this issue may also impact the official java client. Will wait to hear what others there think.

**Comment by jeffwidman** [comment]
> I sent a note to the kafka dev list because I believe this issue may also impact the official java client. Will wait to hear what others there think.

@dpkp was there any further discussion on the dev list about this or JIRA tickets against the Java client?

**Comment by dpkp** [comment]
This issue should be resolved with #948 / #1266 . It is also possible, without using a background-thread heartbeat, to update the client.poll() calls that could be triggered by `KafkaConsumer.position()` or `KafkaConsumer.topics()` to pass `delayed_tasks=False`, which will disable the current AutoCommitTask implementation during poll.

**Comment by dpkp** [comment]
Closing now that #1266 has landed
